pkg/engine/plugins.go

```diff
@@ -43,6 +43,80 @@ const (
 	preparePluginVerboseLog = 8
 )
 
+// PackageSet represents a set of packages.
+type PackageSet map[string]workspace.PackageDescriptor
+
+// Add adds a package to this package set.
+func (p PackageSet) Add(plug workspace.PackageDescriptor) {
+	p[plug.String()] = plug
```

Comment:
What's `String()` defined as for a package descriptor? Is it safe w.r.t. parameters etc.? E.g. if you use the same plugin twice with two parameters.

Reply:
It's defined as the package name and version. We're assuming those are unique.
So the same base plugin twice with different parameters is fine as long as its resulted in two different package names. e.g. terraform-provider random, and terraform-provider aws will work.